community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/ForkedProcessorStep.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.unsafe.impl.batchimport.staging;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.LockSupport;
+
+import static java.util.concurrent.locks.LockSupport.park;
+
+/**
+ * Executes batches sequentially, one at a time, although running multiple processing threads on each batch.
+ * This is almost the opposite of {@link ProcessorStep} which has ability of running multiple batches in parallel,
+ * each batch processed by one thread.
+ *
+ * The purpose of this type of step is to much better be able to parallelize steps that are seen to be
+ * bottlenecks, but are generally very hard to figure out how to parallelize.
+ *
+ * Extending {@link ProcessorStep} and providing max processors 1, i.e. always single-threaded, i.e. only
+ * max one batch being processed at any given point in time. This thread instead starts its own army of
+ * internal "forked" processors which will sit and wait for notifications to start processing the next batch.
+ */
+public abstract class ForkedProcessorStep<T> extends ProcessorStep<T>
+{
+    // used by forked processors to count down when they're done, so that main processing thread
+    // knows when they're all done
+    private final AtomicInteger doneSignal = new AtomicInteger();
+    private final int maxForkedProcessors;
+    protected final List<ForkedProcessor> forkedProcessors = new ArrayList<>();
+    // main processing thread communicates batch to process using this variable
+    // it's not volatile, but piggy-backs on globalTicket for that
+    private T currentBatch;
+    // this ticket helps coordinating with the forked processors
+    private long globalTicket;
```

Comment:
There's a comment in `process` of that class which explains that it isn't needed